src/management-system-v2/components/confirmation-button.tsx

```diff
@@ -0,0 +1,77 @@
+import { ComponentProps, FC, PropsWithChildren, useState } from 'react';
+import { Button, Modal, Tooltip } from 'antd';
+
+type ConfirmationModalProps = {
+  onConfirm: (...args: any[]) => Promise<any> | any;
+  title: string;
+  description: string;
+  canCloseWhileLoading?: boolean;
+  modalProps?: Omit<
+    ComponentProps<typeof Modal>,
+    'open' | 'title' | 'onOk' | 'confirmLoading' | 'onCancel'
+  >;
+  buttonProps?: ComponentProps<typeof Button>;
+  tooltip?: string;
+};
+
+const ConfirmationButton: FC<PropsWithChildren<ConfirmationModalProps>> = ({
+  children,
+  onConfirm,
+  title,
+  description,
+  canCloseWhileLoading = false,
+  modalProps,
+  buttonProps,
+  tooltip,
+}) => {
+  const [modalOpen, setModalOpen] = useState(false);
+  const [loading, setLoading] = useState(false);
+
+  const clearModal = () => {
+    setModalOpen(false);
+    setLoading(false);
+  };
+
+  const onConfirmWrapper = () => {
+    setLoading(true);
+    const possiblePromise = onConfirm();
+
+    if (
+      typeof possiblePromise === 'object' &&
+      'then' in possiblePromise &&
+      typeof possiblePromise.then === 'function'
+    )
+      possiblePromise.then(clearModal).catch(clearModal);
+    else clearModal();
+  };
+
+  return (
+    <>
+      <Modal
+        closeIcon={null}
+        {...modalProps}
+        title={title}
+        open={modalOpen}
+        onOk={onConfirmWrapper}
+        confirmLoading={loading}
+        onCancel={() => (canCloseWhileLoading || !loading) && setModalOpen(false)}
+        cancelButtonProps={{ disabled: loading }}
```

Comment:
Can a cancel while loading happen when the button is disabled while loading? So can onCancel be called while loading?

Reply:
Yes, onCancel also gets called for a click on the mask (the background behind modal).